src/d3a/models/myco_matcher/external_matcher.py

```diff
@@ -1,140 +1,179 @@
+        channel = f"{self.response_channel}/matched-recommendations/"
+        response_data = {"event": "match", "status": "success"}
         data = json.loads(message.get("data"))
         recommendations = data.get("recommended_matches", [])
-        validated_records = {}
-        for record in recommendations:
-            market = self.markets_mapping.get(record.get("market_id"), None)
-            if market is None or market.readonly:
-                # The market is already finished or doesn't exist
-                continue
-
-            bid = offer_or_bid_from_json_string(json.dumps(record.get("bid")))
-            offer = offer_or_bid_from_json_string(json.dumps(record.get("offer")),
-                                                  record.get("bid").get("time"))
-
-            try:
-                validate_authentic_bid_offer_pair(
-                    bid,
-                    offer,
-                    record.get("trade_rate"),
-                    record.get("selected_energy")
-                    )
-
-                if not (offer.id in market.offers and bid.id in market.bids):
-                    # Offer or Bid either don't belong to market or were already matched
-                    raise InvalidBidOfferPair
-
-                if record.get("market_id") not in validated_records:
-                    validated_records[record.get("market_id")] = []
-
-                validated_records[record.get("market_id")].append(BidOfferMatch(
-                    bid,
-                    record.get("selected_energy"),
-                    offer,
-                    record.get("trade_rate")))
-            except InvalidBidOfferPair as ex:
-                # If validation fails or offer/bid were consumed
-                response_dict["status"] = "fail"
-                response_dict["message"] = "Validation Error"
-                logging.exception(f"Bid offer pair validation failed with error {ex}")
-                break
-        if response_dict["status"] == "success":
+        try:
+            validated_records = self._get_validated_bid_offer_match_list(recommendations)
             for market_id, records in validated_records.items():
                 market = self.markets_mapping.get(market_id)
-                if market.readonly:
-                    # The market has just finished
+                try:
+                    market.match_recommendation(records)
+                except (OfferNotFoundException, BidNotFoundException):
+                    # If the offer or bid have just been consumed
                     continue
-                market.match_recommendation(records)
-        self.myco_ext_conn.publish_json(channel, response_dict)
+        except Exception as ex:
+            response_data["status"] = "fail"
+            response_data["message"] = "Validation Error"
+            if not isinstance(ex, MycoValidationException):
+                logging.exception("Bid offer pair matching failed.")
```

Comment:
This message should help us to understand what went wrong. Is it possible to add some more information to help our future selves? Or would any information be useless at this point?

Reply:
atm this statement `if market is None or market.readonly`  is the only place where this exception is raised
anw it is too late to display something at this stage, so I will add some message when raising the exception

Reply:
this is indeed late here, however the exception traceback should be enough